### ISSUE TITLE
DNS leaks when using with worker. 

### ISSUE THREAD

**Issue by timmyvo** [tina-hello/doh-cf-workers#11]
Hi Tina, 
Could you please spend your free time to have a check DNS leaking issue when using this script? 
I notice when I change the DoH endpoint to my nextDNS DoH link. Instead there should only be NextDNS server for queries, it's going to leak to OpenDNS, YandexDNS, google DNS as well. 
I did  use YogaDNS for resolving on my windows. Even setting up fix DoH native on Windows 11. Sometimes, it is still leaking to other servers. 
I suspect the worker.dev domain might be the culprit in this case but I don't have enough technical knowledge to dig deeply into the issue. Hopefully you can have a quick check. 
Notice that I have no DNS leaking when I use directly my nextDNS DoH endpoint with YogaDNS app, not proxy through the workers scripts. 
Thank you for you support. Cheers.

**Comment by tina-hello** [comment]
Interesting, seems to be an oddity in either YogaDNS, Windows 11, or the browser. I tried adding a worker proxy to my personal NextDNS path and then using that address in the Chrome DoH setting, no leak was detected with https://dnscheck.tools/

I do notice that Windows 11 DoH native refuses to only use one DoH endpoint, if you only define one it will randomly resolve with unencrypted queries from time to time, so both primary & secondary must be defined in IPv4 & IPv6. As it happened, yesterday Cloudflare had a glitch that temporarily brought down some servers, it might also affect worker resolution.

How did you test your setup? The website you use, the browser setting etc.

**Comment by timmyvo** [comment]
I did run it through DoH inside Chrome, it is still leaking. Not all time it will leak, you should refresh multiple times for the website  https://dnscheck.tools/, sometimes it leaks, sometime no.  I can share it info with you if you want to verify. You can have a look into attached image for reference. 

![image](https://github.com/tina-hello/doh-cf-workers/assets/26508204/e7fc2c59-3639-447a-9770-e9dbbde240a9)

**Comment by timmyvo** [comment]
Thank you so much for your support. Waiting for your investigation. I remembered there is no leak in the past, I've just noticed the leaks recently.

**Comment by tina-hello** [comment]
@timmyvo I made a repo in https://github.com/tina-hello/workers-leaktest/ to test it, in my several runs with 400 domains on each, everything resolves correctly to the NextDNS block page, which means all answers come from NextDNS, since I checked for `blockpage.nextdns.io` in the result, which shouldn't be available from any other DNS server. So all filtering from NextDNS should still work.

Since I still can replicate the leak detection itself from both Firefox & Chrome (so it's not browser issue, unless both of them somehow use the same library that would send DoH queries to random servers), and no leak is detected when I use NextDNS directly (so it's not a NextDNS issue), and turns out I can trigger the detection on https://browserleaks.com/dns too with similar result (so its not a bug in https://dnscheck.tools/, unless both of them use very similar faulty algorithms), I don't know where to go from here.

Tracking this further would require having a nameserver that logs requests, which I don't have and am not really interested in setting up due to the risk of being abused for amplification attacks.

I'll keep this issue open in case anyone has the infrastructure to debug it themselves.

**Comment by timmyvo** [comment]
Hi Tina,
Thank you for your support and your explanation. Hope someone else can figured out the best way to fix this.

**Comment by tina-hello** [comment]
https://www.reddit.com/r/Adguard/comments/174xnmw/google_public_dns_addresses_when_using_adguard_dns/ noted that browserleaks report AdGuard DNS to leak to Google, while https://help.nextdns.io/t/y4yhg05/nextdns-is-connecting-to-cloudflare noted browserleaks report NextDNS to leak to Cloudflare. I'm less inclined to believe both AdGuard DNS and NextDNS screw up that bad, and more about browserleaks faulty detection.

**Comment by timmyvo** [comment]
Do you think? that somehow cloudflare worker resolves its domain workers.dev via multiple DNS servers so that might be the reason we've got leaking. I did try to deply rethinkDNS on cloudlfare worker. It iwas leaking too. I don't know how to verify this assumption. If you know how to do it, can you please check that. Thank you.

**Comment by tina-hello** [comment]
Nah, the nameserver for workers.dev are controlled by Cloudflare, there's no reason for DNS leak sites to know what DNS is being used to resolve that, they can only see what is being used to resolve their own domain.

**Comment by timmyvo** [comment]
I have thoroughly reviewed the situation with RethinkDNS and I am confident that there is no leakage. So when you have free time, would you be able to take a moment and verify on your end as well. If you find that there are no leaks, we can then narrow down the where issue possibly come from. Thank you for your dedicated support. Hope to get your reply soon. Cheers

**Comment by tina-hello** [comment]
Sorry for the late reply. Do you mean you don't find any leaks on your RethinkDNS instance, or do you mean when you use this repo to forward to RethinkDNS?

**Comment by ManuelGotzen** [comment]
DNS LEAK CHECK LINK WAS HELPFUL THANKS TINA

FOUND MY A MISTAKE. 6 MONTHS GIVEN AWAY. 
HAD ADGUARD-DNS.IO SERVER SETTING 

IP6: 2a10:50c0::ad1:ff. 
THIS WAS **CHANGED** TO 
IP6: 2a10:50c0:**0:0:0:0**:ad1:ff 


I'm so lost in the settings and user scripts that I thought I had caused the error. I THINK MY LOCAL UP AND DOWNSTREAM CONNECTION IN THE OWN AWS AND OWN CLOUDWARE DASH in conjunction with a DNS WILDCARD HOST LIST HAVE ACHIEVED THIS. 

Yeah. ADHD users usually only read after they have used it. ERGO my userscript made the ULTIMATE list cause the errors. 

> I JUST WANTED AN ADVERTISING-FREE FireTV (WITHOUT ROOT AND WITHOUT ROUTER ACCESS! WITHOUT ADGUARD HOME!!!) THAT'S TRICKY!!! CURRENTLY I AND Amazon's 3rd level support can't get any further here. 😤🤯_Need real help. 🐇 Urgent_. 

Well,
Who else besides me comes up with the idea of ​​alienating the device backup system in the userscript via the googleapis by changing the system language for the purpose of changing the vendor cookie settings for the who for the advertising settings?! Lol Me! That worked well too. Until Amazon carried out an update at the same time. TABOS and FIREOS were swapped at this point due to. and also in the logged in child profile in the sales demo. I didn't want to make it too easy for myself. LOL... Fix this... I don't have a solution yet. All I know is that if I turn off the devices, not even the factory settings will save me. ... That much is certain at this point. 

I miss my old DELL SERVER WITH TRIPLE FIREWALL. 😭

Sorry, didn't mean to steal your time. 
(But I'm really desperate. And tired) 

(~￣³￣)~ 


 , 
g 

/ Manuel Gotzen 

**THX** &
HAVE **A**
**GOOD**
**TIME**

__
send via
AMAZON KINDLE FIRE HD10 (KFMUWI)
